Managers/PerformanceMan.cpp

```diff
@@ -0,0 +1,264 @@
+
+	void PerformanceMan::StartPerformanceMeasurement(PerformanceCounters counter) { m_PerfMeasureStart[counter] = g_TimerMan.GetAbsoulteTime(); }
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void PerformanceMan::StopPerformanceMeasurement(PerformanceCounters counter) {
+		m_PerfMeasureStop[counter] = g_TimerMan.GetAbsoulteTime();
+		AddPerformanceSample(counter, m_PerfMeasureStop[counter] - m_PerfMeasureStart[counter]);
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void PerformanceMan::NewPerformanceSample() {
+		m_Sample++;
+		if (m_Sample >= c_MaxSamples) { m_Sample = 0; }
+
+		for (unsigned short pc = 0; pc < PERF_COUNT; ++pc) {
+			m_PerfData[pc][m_Sample] = 0;
+			m_PerfPercentages[pc][m_Sample] = 0;
+		}
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void PerformanceMan::Update() {
+		// Time and add the millisecs per frame reading to the buffer
+		m_MSPFs.push_back(m_FrameTimer->GetElapsedRealTimeMS());
+		// Keep the buffer trimmed
+		while (m_MSPFs.size() > c_MSPFAverageSampleSize) {
+			m_MSPFs.pop_front();
+		}
+		// Calculate the average milliseconds per frame over the last sampleSize frames
+		m_MSPFAverage = 0;
+		for (std::deque<unsigned int>::iterator fItr = m_MSPFs.begin(); fItr != m_MSPFs.end(); ++fItr) {
+			m_MSPFAverage += *fItr;
+		}
+		m_MSPFAverage /= m_MSPFs.size();
+
+		// TODO: This probably belongs in TimerMan so figure out where exactly to shove it in it.
+		// If one sim update per frame mode, adjust the pitch of most sound effects to match the sim time over real time ratio as it fluctuates!
+		if (g_TimerMan.IsOneSimUpdatePerFrame()) {
+			// Calculate the sim speed over the actual real time
+			m_SimSpeed = g_TimerMan.GetDeltaTimeMS() / static_cast<float>(m_MSPFAverage);
+
+			// If limited, only allow pitch to go slower, not faster
+			if (g_TimerMan.IsSimSpeedLimited() && m_SimSpeed > 1.0) { m_SimSpeed = 1.0; }
+
+			// Soften the ratio of the pitch adjustment so it's not such an extreme effect on the audio
+			// TODO: Don't hardcode this coefficient - although it's a good default
+			float pitch = m_SimSpeed + (1.0F - m_SimSpeed) * 0.35;
+			// Set the pitch for all other applicable sounds other than music
+			g_AudioMan.SetGlobalPitch(pitch, true);
+		} else {
+			m_SimSpeed = 1.0;
+		}
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void PerformanceMan::Draw(AllegroBitmap bitmapToDrawTo) {
+		if (m_ShowPerfStats) {
+			// Time and add the millisecs per frame reading to the buffer
+			m_MSPFs.push_back(m_FrameTimer->GetElapsedRealTimeMS());
```

Comment:
I don't understand why we have all this saving MSPFs stuff here as well as in Update. Also, I don't understand why we don't reset m_MSPFAverage below like we do in Update.

Reply:
I don't either, but I do understand that this somehow evens out the fluctuations in fps.
I was messing around with removing either of these and at one point ended up with ~70 fps in the debug build...

Reply:
Ummm, shit, okay. Leave it be then but drop a TODO to investigate it sometime so we know it's something that seems weird but we don't understand.